pkg/syncer/k8scloudoperator/placement.go

```diff
@@ -410,6 +416,170 @@ func stampPVCWithError(ctx context.Context, client kubernetes.Interface,
+// Sibling PVCs are the ones that belong to the same replica as that of currPVC.
+func getCousinPVCs(ctx context.Context, client kubernetes.Interface, currPVC *v1.PersistentVolumeClaim) ([]string, error) {
+	log := logger.GetLogger(ctx)
+
+	// Find the statefulset that the PVC belongs to
+	parentSTS, err := getParentSTS(ctx, client, currPVC)
+	if err != nil {
+		return []string{}, err
+	}
+	// parent statefulset not found, send back empty list of cousin PVCs
+	if parentSTS == nil {
+		return []string{}, nil
+	}
+
+	cousinPVCs := []string{}
+	// Number of replicas defined in the statefulset
+	numberOfReplicas := int(*parentSTS.Spec.Replicas)
+
+	// Find the replica number that currPVC belongs to.
+	currPVCParentReplica, err := getPVCParentReplicaID(ctx, currPVC.Name)
+	if err != nil {
+		log.Errorf("Failed to get replica number to which the PVC belongs. Err %+v", err)
+		return []string{}, err
+	}
+
+	for _, volume := range parentSTS.Spec.VolumeClaimTemplates {
+		for i := 0; i < numberOfReplicas; i++ {
+			// If replica under consideration is same as the one that currPVC belongs to, ignore it as it will give sibling PVCs.
+			if i == currPVCParentReplica {
+				continue
+			}
+			// Construct the cousin PVC name and add it to the list
+			cousinPVCName := volume.ObjectMeta.Name + "-" + parentSTS.Name + "-" + strconv.Itoa(i)
+			cousinPVCs = append(cousinPVCs, cousinPVCName)
+		}
+	}
+
+	return cousinPVCs, nil
+}
+
+// Removes host at the given index
+func filterHost(s []string, index int) []string {
+	return append(s[:index], s[index+1:]...)
+}
+
+// eliminateNodesWithPvcOfSiblingReplica filters out the nodes that have bound PVCs of sibling replicas.
+// It finds cousin PVCs of currPVC and elimates all nodes which have at least one cousin PVC placed on them.
+func eliminateNodesWithPvcOfSiblingReplica(ctx context.Context, client kubernetes.Interface, currPVC *v1.PersistentVolumeClaim, candidateHosts []string) ([]string, error) {
+	log := logger.GetLogger(ctx)
+
+	// Proceed only if it is a vDPP PVC
+	pvcLabels := currPVC.GetLabels()
+	if _, ok := pvcLabels[appplatformLabel]; !ok {
+		log.Infof("Not a vDPP PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	// Do not proceed if it is a standalone pod
+	pvcAnnotation := currPVC.ObjectMeta.Annotations
+	if _, ok := pvcAnnotation[nodeAffinityAnnotationKey]; ok {
+		log.Infof("Node affinity key already specified for PVC %+v, skip this step.", currPVC.Name)
+		return candidateHosts, nil
+	}
+
+	cousinPVCs, err := getCousinPVCs(ctx, client, currPVC)
+	if err != nil {
+		return []string{}, err
+	}
+
+	for _, cousinPVC := range cousinPVCs {
+		pvc, err := client.CoreV1().PersistentVolumeClaims(currPVC.Namespace).Get(ctx, cousinPVC, metav1.GetOptions{})
+		if err != nil {
+			return []string{}, err
+		}
+
+		pvName := pvc.Spec.VolumeName
+		pv, err := client.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
```

Comment:
We can avoid this api call by using volume.kubernetes.io/selected-node for most of the PVCs. It is very rare that this annotation will be missing in PVCs, if it is present it is always correct.

We can optimize the code by checking for volume.kubernetes.io/selected-node annotation if it is not present then we can get PV. This will reduce the number of api-server calls to 2(one to fetch all StatefulSets in a namespace and other to fetch all PVCs of an instance) per PVC placement.

Reply:
It is possible that for some of the PVCs, volumes may not be created yet. We should ignore such volumes.

Reply:
We can avoid this api call by using volume.kubernetes.io/selected-node for most of the PVCs. It is very rare that this annotation will be missing in PVCs, if it is present it is always correct.

We can optimize the code by checking for volume.kubernetes.io/selected-node annotation if it is not present then we can get PV. This will reduce the number of api-server calls to 2(one to fetch all StatefulSets in a namespace and other to fetch all PVCs of an instance) per PVC placement.